### PR TITLE
Add default value when generate the update rows

### PR DESCRIPTION
**Pull request by july2993**
### What is changed and how it works?
only on the condition PreWrite.schemaVersion < schemaVersion of the
table replicated to downstream.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

enable amend:
`
set global tidb_enable_amend_pessimistic_txn = 1;
`

run:
```
/* INIT */ drop table if exists t;
/* INIT */ create table t (id int primary key, c_str varchar(20));
/* INIT */ insert into t values (1, '0001'), (2, '0002'), (3, null), (4, '0003'), (5, null);
/* TXN */ begin;
/* TXN */ insert into t values (6, '0004');
/* TXN */ insert into t values (7, null);
/* DDL */ alter table t add c_str_new varchar(20);   # IN ANOTHER SESSION
/* TXN */ update t set c_str = '0005' where id = 1;
/* TXN */ update t set c_str = null where id = 2;
/* TXN */ update t set c_str = '0006' where id = 3;
/* TXN */ delete from t where id = 4;
/* TXN */ delete from t where id = 5;
/* TXN */ commit;
/* TXN */ select * from t order by id;
/* TXN */ admin check table t;

```

Code changes

 
Side effects

### Release note
- No Release note.

### CODE REVIEW

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by you06** [comment]
Tested with this PR got unexpected NULL value in down stream
```
MySQL [test]> select * from accounts;                                                          │MySQL [test]> select * from accounts;
+----+---------+------+------+                                                                 │+----+---------+------+------+
| id | balance | aa   | bb   |                                                                 │| id | balance | aa   | bb   |
+----+---------+------+------+                                                                 │+----+---------+------+------+
|  1 |       1 |    1 |   10 |                                                                 │|  1 |       1 |    1 | NULL |
|  2 |       2 |    2 |   10 |                                                                 │|  2 |       2 |    2 | NULL |
|  3 |       3 |    3 |   10 |                                                                 │|  3 |       3 |    3 | NULL |
+----+---------+------+------+                                                                 │+----+---------+------+------+
3 rows in set (0.063 sec)                                                                      │3 rows in set (0.098 sec)
                                                                                               │
MySQL [test]> show create table accounts;                                                      │MySQL [test]> show create table accounts;
+----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
------------------------------------------------+                                              │--------------------------------------------------+
| Table    | Create Table                                                                      │| Table    | Create Table                                                                     
                                                                                               │                                                                                              
                                                |                                              │                                                  |
+----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
------------------------------------------------+                                              │--------------------------------------------------+
| accounts | CREATE TABLE `accounts` (                                                         │| accounts | CREATE TABLE `accounts` (
  `id` int(11) NOT NULL,                                                                       │  `id` int(11) NOT NULL,
  `balance` bigint(20) DEFAULT NULL,                                                           │  `balance` bigint(20) DEFAULT NULL,
  `aa` int(11) DEFAULT NULL,                                                                   │  `aa` int(11) DEFAULT NULL,
  `bb` int(11) DEFAULT 10,                                                                     │  `bb` int(11) DEFAULT 10,
  PRIMARY KEY (`id`)                                                                           │  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |                                  │) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
------------------------------------------------+                                              │--------------------------------------------------+
1 row in set (0.051 sec)                                                                       │1 row in set (0.124 sec)
                                                                                               │
```

**Comment by july2993** [comment]
> R got unexpecte

can you describe the detailed step?

**Comment by july2993** [comment]
> Tested with this PR got unexpected NULL value in down stream
> 
> ```
> MySQL [test]> select * from accounts;                                                          │MySQL [test]> select * from accounts;
> +----+---------+------+------+                                                                 │+----+---------+------+------+
> | id | balance | aa   | bb   |                                                                 │| id | balance | aa   | bb   |
> +----+---------+------+------+                                                                 │+----+---------+------+------+
> |  1 |       1 |    1 |   10 |                                                                 │|  1 |       1 |    1 | NULL |
> |  2 |       2 |    2 |   10 |                                                                 │|  2 |       2 |    2 | NULL |
> |  3 |       3 |    3 |   10 |                                                                 │|  3 |       3 |    3 | NULL |
> +----+---------+------+------+                                                                 │+----+---------+------+------+
> 3 rows in set (0.063 sec)                                                                      │3 rows in set (0.098 sec)
>                                                                                                │
> MySQL [test]> show create table accounts;                                                      │MySQL [test]> show create table accounts;
> +----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
> -----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
> ------------------------------------------------+                                              │--------------------------------------------------+
> | Table    | Create Table                                                                      │| Table    | Create Table                                                                     
>                                                                                                │                                                                                              
>                                                 |                                              │                                                  |
> +----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
> -----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
> ------------------------------------------------+                                              │--------------------------------------------------+
> | accounts | CREATE TABLE `accounts` (                                                         │| accounts | CREATE TABLE `accounts` (
>   `id` int(11) NOT NULL,                                                                       │  `id` int(11) NOT NULL,
>   `balance` bigint(20) DEFAULT NULL,                                                           │  `balance` bigint(20) DEFAULT NULL,
>   `aa` int(11) DEFAULT NULL,                                                                   │  `aa` int(11) DEFAULT NULL,
>   `bb` int(11) DEFAULT 10,                                                                     │  `bb` int(11) DEFAULT 10,
>   PRIMARY KEY (`id`)                                                                           │  PRIMARY KEY (`id`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |                                  │) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
> +----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
> -----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
> ------------------------------------------------+                                              │--------------------------------------------------+
> 1 row in set (0.051 sec)                                                                       │1 row in set (0.124 sec)
>                                                                                                │
> ```

can you describe the detailed step?

**Comment by you06** [comment]
@july2993  I ran the following SQLs from upstream

```sql
/* init */ drop table if exists accounts;
/* init */ create table accounts(id int primary key, balance bigint, aa int);
/* init */ insert into accounts values(1, 100, 1), (2, 0, 2), (3, 1, 3);

/* TXN1 */ set session tidb_enable_amend_pessimistic_txn = 1;
/* DDL */ set session tidb_enable_amend_pessimistic_txn = 1;
/* TXN1 */ begin;
/* DDL */ alter table accounts add column bb int default 10;
/* TXN1 */ update accounts set balance = id where 1;
/* TXN1 */ commit;
/* TXN1 */ select * from accounts;
```

File: drainer/translator/mysql.go [hunk 1]
```diff
@@ -56,9 +56,9 @@ func genMysqlInsert(schema string, table *model.TableInfo, row []byte) (names []
 	return names, args, nil
 }
 
-func genMysqlUpdate(schema string, table *model.TableInfo, row []byte, isTblDroppingCol bool) (names []string, values []interface{}, oldValues []interface{}, err error) {
+func genMysqlUpdate(schema string, table *model.TableInfo, row []byte, canAppendDefaultValue bool) (names []string, values []interface{}, oldValues []interface{}, err error) {
 	columns := writableColumns(table)
-	updtDecoder := newUpdateDecoder(table, isTblDroppingCol)
+	updtDecoder := newUpdateDecoder(table, canAppendDefaultValue)
```
Comment (you06) [inline drainer/translator/mysql.go:61]: The function sign of `newUpdateDecoder` is still `(table *model.TableInfo, isTblDroppingCol bool) updateDecoder`.

**Review by lichunzhu — COMMENTED** [review]
Should we add an integration test for this case?

File: drainer/translator/translator.go [hunk 2]
```diff
@@ -158,11 +160,11 @@ func DecodeOldAndNewRow(b []byte, cols map[int64]*model.ColumnInfo, loc *time.Lo
 	parsedCols := cnt / 2
 	isInvalid := len(newRow) != len(oldRow) || (len(cols) != parsedCols && len(cols)-1 != parsedCols)
 	if isInvalid {
-		return nil, nil, errors.Errorf("row data is corrupted %v", b)
+		return nil, nil, errors.Errorf("row data is corrupted cols num: %d, oldRow: %v, newRow: %v", len(cols), oldRow, newRow)
 	}
 	if parsedCols == len(cols)-1 {
```
Comment (you06) [inline drainer/translator/translator.go:165]: It seems there may be multi columns changed, I tested this sample and drainer quit with error.

```sql
/* init */ drop table if exists accounts;
/* init */ create table accounts(id int primary key, balance bigint, aa int);
/* init */ insert into accounts values(1, 100, 1), (2, 0, 2), (3, 1, 3);

/* TXN1 */ set session tidb_enable_amend_pessimistic_txn = 1;
/* DDL */ set session tidb_enable_amend_pessimistic_txn = 1;
/* TXN1 */ begin;
/* DDL */ alter table accounts add column bb int default 10;
/* DDL */ alter table accounts add column cc int default 20;
/* TXN1 */ update accounts set balance = id where 1;
/* TXN1 */ commit;
/* TXN1 */ select * from accounts;
```

```
[2020/11/03 07:58:30.335 +00:00] [ERROR] [server.go:289] ["syncer exited abnormal"] [error="failed to add item: gen update fail: table `test`.`accounts`: row data is corrupted cols num: 5, oldRow: map[1:KindInt64 1 2:KindInt64 100 3:KindInt64 1], newRow: map[1:KindInt64 1 2:KindInt64 1 3:KindInt64 1]"] [errorVerbose="row data is corrupted cols num: 5, oldRow: map[1:KindInt64 1 2:KindInt64 100 3:KindInt64 1], newRow: map[1:KindInt64 1 2:KindInt64 1 3:KindInt64 1]\ngithub.com/pingcap/tidb-binlog/drainer/translator.DecodeOldAndNewRow\n\t/home/tidb/issue-20500/tidb-binlog/drainer/translator/translator.go:163\ngithub.com/pingcap/tidb-binlog/drainer/translator.updateDecoder.decode\n\t/home/tidb/issue-20500/tidb-binlog/drainer/translator/translator.go:211\ngithub.com/pingcap/tidb-binlog/drainer/translator.genMysqlUpdate\n\t/home/tidb/issue-20500/tidb-binlog/drainer/translator/mysql.go:65\ngithub.com/pingcap/tidb-binlog/drainer/translator.TiBinlogToTxn\n\t/home/tidb/issue-20500/tidb-binlog/drainer/translator/mysql.go:159\ngithub.com/pingcap/tidb-binlog/drainer/sync.(*MysqlSyncer).Sync\n\t/home/tidb/issue-20500/tidb-binlog/drainer/sync/mysql.go:186\ngithub.com/pingcap/tidb-binlog/drainer.(*Syncer).run\n\t/home/tidb/issue-20500/tidb-binlog/drainer/syncer.go:378\ngithub.com/pingcap/tidb-binlog/drainer.(*Syncer).Start\n\t/home/tidb/issue-20500/tidb-binlog/drainer/syncer.go:132\ngithub.com/pingcap/tidb-binlog/drainer.(*Server).Start.func4\n\t/home/tidb/issue-20500/tidb-binlog/drainer/server.go:288\ngithub.com/pingcap/tidb-binlog/drainer.(*taskGroup).start.func1\n\t/home/tidb/issue-20500/tidb-binlog/drainer/util.go:75\nruntime.goexit\n\t/usr/local/share/go/src/runtime/asm_amd64.s:1374\ntable `test`.`accounts`\ngen update fail\nfailed to add item"]
```

**Comment by you06** [comment]
> Tested with this PR got unexpected NULL value in down stream
> 
> ```
> MySQL [test]> select * from accounts;                                                          │MySQL [test]> select * from accounts;
> +----+---------+------+------+                                                                 │+----+---------+------+------+
> | id | balance | aa   | bb   |                                                                 │| id | balance | aa   | bb   |
> +----+---------+------+------+                                                                 │+----+---------+------+------+
> |  1 |       1 |    1 |   10 |                                                                 │|  1 |       1 |    1 | NULL |
> |  2 |       2 |    2 |   10 |                                                                 │|  2 |       2 |    2 | NULL |
> |  3 |       3 |    3 |   10 |                                                                 │|  3 |       3 |    3 | NULL |
> +----+---------+------+------+                                                                 │+----+---------+------+------+
> 3 rows in set (0.063 sec)                                                                      │3 rows in set (0.098 sec)
>                                                                                                │
> MySQL [test]> show create table accounts;                                                      │MySQL [test]> show create table accounts;
> +----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
> -----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
> ------------------------------------------------+                                              │--------------------------------------------------+
> | Table    | Create Table                                                                      │| Table    | Create Table                                                                     
>                                                                                                │                                                                                              
>                                                 |                                              │                                                  |
> +----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
> -----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
> ------------------------------------------------+                                              │--------------------------------------------------+
> | accounts | CREATE TABLE `accounts` (                                                         │| accounts | CREATE TABLE `accounts` (
>   `id` int(11) NOT NULL,                                                                       │  `id` int(11) NOT NULL,
>   `balance` bigint(20) DEFAULT NULL,                                                           │  `balance` bigint(20) DEFAULT NULL,
>   `aa` int(11) DEFAULT NULL,                                                                   │  `aa` int(11) DEFAULT NULL,
>   `bb` int(11) DEFAULT 10,                                                                     │  `bb` int(11) DEFAULT 10,
>   PRIMARY KEY (`id`)                                                                           │  PRIMARY KEY (`id`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |                                  │) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
> +----------+-----------------------------------------------------------------------------------│+----------+----------------------------------------------------------------------------------
> -----------------------------------------------------------------------------------------------│----------------------------------------------------------------------------------------------
> ------------------------------------------------+                                              │--------------------------------------------------+
> 1 row in set (0.051 sec)                                                                       │1 row in set (0.124 sec)
>                                                                                                │
> ```

It's caused by `getDefaultOrZeroValue` function will not populate default value column.

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by july2993** [comment]
> Should we add an integration test for this case?

yes, current test using a temporal tool https://github.com/july2993/tk/blob/4465f27496e7c6d8c509e8fde75bfce070c8c4b3/cmd/amend/amend.go#L42
we can port it if this solution if fine

**Comment by you06** [comment]
I tried this case and got error
```sql
/* init */ drop table if exists accounts;
/* init */ create table accounts(id int primary key, balance bigint);
/* init */ insert into accounts values(1, 100), (2, 0), (3, 1);

/* TXN */ set session tidb_enable_amend_pessimistic_txn = 1;
/* DDL */ set session tidb_enable_amend_pessimistic_txn = 1;
/* TXN */ begin;
/* DDL */ alter table accounts add column bb int not null;
/* DDL */ alter table accounts add column cc int not null;
/* TXN */ update accounts set balance = id where 1;
/* TXN */ commit;
/* TXN */ select * from accounts;
```

```log
[2020/11/04 08:08:34.838 +00:00] [ERROR] [executor.go:111] ["Exec fail, will rollback"] [query="REPLACE INTO `test`.`accounts`(`balance`,`bb`,`id`) VALUES(?,?,?)"] [args="[3,0,3]"] [error="Error 1364: Field 'cc' doesn't have a default value"]
```

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

File: drainer/schema.go [hunk 3]
```diff
@@ -173,8 +185,33 @@ func (s *Schema) DropTable(id int64) (string, error) {
 	return table.Name.O, nil
 }
 
+func (s *Schema) appendTableInfo(schemaVersion int64, table *model.TableInfo) {
+	tbls := s.tables[table.ID]
+	tbls = append(tbls, schemaVersionTableInfo{SchemaVersion: schemaVersion, TableInfo: table})
+	if len(tbls) > 2 {
+		tbls = tbls[len(tbls)-2:]
```
Comment (csuzhangxc) [inline drainer/schema.go:192]: Does _amend_pessimistic_txn_ add only one additional schema version at most?

Comment (july2993) [reply drainer/schema.go:192]: um..I think we can save the limit version number in a time range like 1hour or 1day?

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Review by csuzhangxc — COMMENTED** [review]
LGTM

**Comment by csuzhangxc** [comment]
@you06 PTAL again?

**Comment by you06** [comment]
I'm running some tests on the latest commit.

**Comment by zyguan** [comment]
I did some tests today, there is a problem with enum type. The case is
```sql
/* init */ drop table if exists t;
/* init */ create table t (id int primary key, c int);
/* init */ insert into t (id, c) values (1, 1);

/* txn */ begin;
/* txn */ update t set c = 2 where id = 1;
/* ddl */ alter table t add column cc enum('a','b','c','d','e') default 'b';
/* txn */ commit;
```

Drainer will execute `REPLACE INTO test.t(c, cc, id) VALUES(2,0,1)` on downstream, which will be rejected with `Error 1265: Data truncated for column 'cc' `. Drainer retried the txn for several times and finally exit with error. That is, we got wrong default value for this enum column (`Datum.b` is "b" but `Datum.i` is 0).

**Comment by july2993** [comment]
> I did some tests today, there is a problem with enum type. The test case is
> 
> ```sql
> /* init */ drop table if exists t;
> /* init */ create table t (id int primary key, c int);
> /* init */ insert into t (id, c) values (1, 1);
> 
> /* txn */ begin;
> /* txn */ update t set c = 2 where id = 1;
> /* ddl */ alter table t add column cc enum('a','b','c','d','e') default 'b';
> /* txn */ commit;
> ```
> 
> Drainer will execute `REPLACE INTO test.t(c, cc, id) VALUES(2,0,1)` on downstream, which will be rejected with `Error 1265: Data truncated for column 'cc' `. Drainer retried the txn for several times and finally exit with error. That is, we got a wrong default value for this enum column (`Datum.b` is "b" but `Datum.i` is 0).

fix in https://github.com/pingcap/tidb-binlog/pull/1006/commits/18393fe97a8f1e440dc775fc92263676a7494a90

File: drainer/translator/table_info.go [hunk 4]
```diff
@@ -19,5 +19,7 @@ import "github.com/pingcap/parser/model"
 type TableInfoGetter interface {
 	TableByID(id int64) (info *model.TableInfo, ok bool)
 	SchemaAndTableName(id int64) (string, string, bool)
-	IsDroppingColumn(id int64) bool
+	CanAppendDefaultValue(id int64, schemaVersion int64) bool
+	// IsDroppingColumn(id int64) bool
+	TableBySchemaVersion(version int64, tid int64) (info *model.TableInfo, ok bool)
```
Comment (you06) [inline drainer/translator/table_info.go:24]: The variable name here is a bit confusing.
- `id` and `tid` are both table's id.
- `version` and `schemaVersion` are both schema's version.
- the order of id and version is not same.

**Comment by you06** [comment]
LGTM

**Comment by july2993** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**